core/domain/state_domain.py

```diff
@@ -353,192 +406,1900 @@ def __init__(
+            for ca_name in self.customization_args:
+                customization_args_dict[ca_name] = (
+                    self.customization_args[
+                        ca_name].to_customization_arg_dict()
+                )
+
+        # Consistent with other usages of to_dict() across the codebase, all
+        # values below are plain Python data structures and not domain objects,
+        # despite the names of the keys. This applies to customization_args_dict
+        # below.
+        return {
+            'id': self.id,
+            'customization_args': customization_args_dict,
+            'answer_groups': [group.to_dict() for group in self.answer_groups],
+            'default_outcome': (
+                self.default_outcome.to_dict()
+                if self.default_outcome is not None
+                else None),
+            'confirmed_unclassified_answers': (
+                self.confirmed_unclassified_answers),
+            'hints': [hint.to_dict() for hint in self.hints],
+            'solution': self.solution.to_dict() if self.solution else None,
+        }
+
+    @classmethod
+    def from_dict(cls, interaction_dict):
+        """Return a InteractionInstance domain object from a dict.
+
+        Args:
+            interaction_dict: dict. The dict representation of
+                InteractionInstance object.
+
+        Returns:
+            InteractionInstance. The corresponding InteractionInstance domain
+            object.
+        """
+        default_outcome_dict = (
+            Outcome.from_dict(interaction_dict['default_outcome'])
+            if interaction_dict['default_outcome'] is not None else None)
+        solution_dict = (
+            Solution.from_dict(
+                interaction_dict['id'], interaction_dict['solution'])
+            if (interaction_dict['solution'] and interaction_dict['id'])
+            else None)
+
+        customization_args = (
+            InteractionInstance
+            .convert_customization_args_dict_to_customization_args(
+                interaction_dict['id'],
+                interaction_dict['customization_args']
+            )
+        )
+
+        return cls(
+            interaction_dict['id'],
+            customization_args,
+            [AnswerGroup.from_dict(h)
+             for h in interaction_dict['answer_groups']],
+            default_outcome_dict,
+            interaction_dict['confirmed_unclassified_answers'],
+            [Hint.from_dict(h) for h in interaction_dict['hints']],
+            solution_dict)
+
+    def to_proto(self):
+        """Returns a proto representation of the interaction object.
+
+        Returns:
+            InteractionInstance. The InteractionInstance proto object.
+        """
+        interaction_proto = None
+        if self.id == 'Continue':
+            interaction_proto = state_pb2.InteractionInstance(
+                continue_instance=self._to_continue_proto(self))
+
+        if self.id == 'FractionInput':
```

Comment:
These should be elif's.

There will also be some interaction IDs that don't have an associated proto (notably the ones not associated with curated lessons). Do you need an "else raise exception" here?

Finally -- you might want to try and make this more declarative. E.g. have a mapping here where the keys are all the interaction IDs in android_validation_constants.py to the corresponding to_proto() method. And add a backend test to make sure the keys in this dict match the IDs in android_validation_constants.py. That way, if a new interaction is added, people will remember to update this since the tests would fail otherwise.

Reply:
> have a mapping here where the keys are all the interaction IDs in android_validation_constants.py to the corresponding to_proto() method

Currently, our proto API doesn't support all the explorations mentioned at `android_validation_constants.py`. Till then we have to depend on this if elif else structure, later we can switch as mentioned by you.